File: overload.ts
// Include unused members to they aren't structurally similar.
class A {a = 0}
class B {b = 0}

function say(a: A, b: B): void;
function say(b: B, a: A): void;
function say(x: A | B, y: A | B) {
  if (x instanceof A && y instanceof B) {
    console.log("ab");
  } else if (x instanceof B && y instanceof A) {
    console.log("ba");
  } else {
    throw new Error("Fail!");
  }
}

function saySwaps(a: A, b: B): void;
function saySwaps(b: B, a: A): void;
function saySwaps(x: A | B, y: A | B) {
  // This is a lie, but it works.
  say(x as A, y as B);
  say(y as B, x as A);
}

function main() {
  let a = new A();
  let b = new B();
  say(a, b);
  // saySwaps(a, b);
  // saySwaps(b, a);
}

main()

File: overload.jl
struct A end
struct B end

function say(a1::A, a2::A)
    println("aa")
end

function say(a::A, b::B)
    println("ab")
end

function say(b::B, a::A)
    println("ba")
end

function say(b1::B, b2::B)
    println("bb")
end

# function say(a, b)
#     println("?")
# end

# function say_swaps(x::Union{A, B}, y::Union{A, B})
# function say_swaps(x::Any, y::Any)
function say_swaps(x, y)
    say(x, y)
    say(y, x)
end

function main()
    a = A()
    b = B()
    say(a, b)
    # say(a, b=B())
    # say_swaps(a, b)
    # say_swaps(b, a)
    # say_swaps(a, 2)
end

main()
